src/core/api/media_element_track_choice_manager.ts

```diff
@@ -0,0 +1,471 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { BehaviorSubject } from "rxjs";
+import EventEmitter from "../../utils/event_emitter";
+import normalizeLanguage from "../../utils/languages";
+import {
+  IAudioTrackPreference,
+  ITextTrackPreference,
+  ITMAudioTrack,
+  ITMAudioTrackListItem,
+  ITMTextTrack,
+  ITMTextTrackListItem,
+  ITMVideoTrack,
+  ITMVideoTrackListItem,
+} from "./track_choice_manager";
+
+// TODO w3c defines an onremovetrack attribute which is not present on
+// ts type definition
+export interface ICustomTextTrackList extends TextTrackList {
+  onremovetrack: ((ev: TrackEvent) => void) | null;
+  onchange: (() => void) | null;
+}
+
+interface IMediaElementTrackChoiceManagerEvents {
+  availableTracksChange: "audio"|"video"|"text";
+  trackChange:
+    { type: "audio"; track: ITMAudioTrack|null } |
+    { type: "video"; track: ITMVideoTrack|null } |
+    { type: "text"; track: ITMTextTrack|null };
+}
+
+/**
+ * Manage video, audio and text tracks for current direct file content.
+ * @class MediaElementTrackChoiceManager
+ */
+export default class MediaElementTrackChoiceManager
+  extends EventEmitter<IMediaElementTrackChoiceManagerEvents> {
+  // Array of preferred languages for audio tracks.
+  // Sorted by order of preference descending.
+  private _preferredAudioTracks : BehaviorSubject<IAudioTrackPreference[]>;
+
+  // Array of preferred languages for text tracks.
+  // Sorted by order of preference descending.
+  private _preferredTextTracks : BehaviorSubject<ITextTrackPreference[]>;
+
+  private _audioTracks : Array<{ track: ITMAudioTrack; origTrack: AudioTrack }>;
+  private _textTracks : Array<{ track: ITMTextTrack; origTrack: TextTrack }>;
+  private _videoTracks : Array<{ track: ITMVideoTrack; origTrack: VideoTrack }>;
+
+  private _mediaElement : HTMLMediaElement;
+
+  constructor(
+    defaults : { preferredAudioTracks : BehaviorSubject<IAudioTrackPreference[]>;
+                 preferredTextTracks : BehaviorSubject<ITextTrackPreference[]>; },
+    mediaElement: HTMLMediaElement
+  ) {
+    super();
+    const { preferredAudioTracks, preferredTextTracks } = defaults;
+
+    this._preferredAudioTracks = preferredAudioTracks;
+    this._preferredTextTracks = preferredTextTracks;
+
+    this._mediaElement = mediaElement;
+
+    /**
+     * Check if track array is different from an other one
+     * @param {Array.<Object>} oldTrackArray
+     * @param {Array.<Object>} newTrackArray
+     * @returns {boolean}
+     */
+    function areTrackArraysDifferent(
+      oldTrackArray: Array<{ origTrack: VideoTrack|AudioTrack|TextTrack }>,
+      newTrackArray: Array<{ origTrack: VideoTrack|AudioTrack|TextTrack }>
+    ) {
```

Comment:
You forgot to add the return type

Reply:
Also this function can be outside the class

Reply:
done